### PR TITLE
Fix dimension hashing bug

### PR DESCRIPTION
**Pull request by gjoranv**
FYI: @hakonhall

### CODE REVIEW

**Review by hakonhall — APPROVED** [review]

File: metrics-proxy/src/main/java/ai/vespa/metricsproxy/service/MetricsParser.java [hunk 1]
```diff
@@ -142,25 +143,42 @@ static private void handleValue(JsonParser parser, Instant timestamp, Consumer c
...
+                                                        dim -> toDimensionId(dim.id), dim -> dim.value)));
+    }
+
+    static long dimensionsHashCode(List<Dimension> dimensions) {
+        int keyHash = 0;
+        int valueHash = 0;
+
+        for (Dimension dim : dimensions) {
+            keyHash += dim.id.hashCode();
+            valueHash += dim.value.hashCode();
```
Comment (baldersheim) [inline metrics-proxy/src/main/java/ai/vespa/metricsproxy/service/MetricsParser.java:172]: Very good catch @gjoranv. This was way to naive.
We can not treat key and value separately.
I suggest using com.yahoo.compress.Hasher.xxh3 and just keep one 64bit.
Either that or follow @hakonhall's suggestion.

File: metrics-proxy/src/test/java/ai/vespa/metricsproxy/service/MetricsParserTest.java [hunk 2]
```diff
@@ -24,6 +25,21 @@ public void consume(Metric metric) {
...
+        var dimensions1 = List.of(
+                new MetricsParser.Dimension("cluster", "CLUSTER-1"),
+                new MetricsParser.Dimension("clusterid", "CLUSTER-1"));
+
+        var dimensions2 = List.of(
+                new MetricsParser.Dimension("cluster", "CLUSTER-2"),
+                new MetricsParser.Dimension("clusterid", "CLUSTER-2"));
+
+        System.out.println(dimensionsHashCode(dimensions1));
+        System.out.println(dimensionsHashCode(dimensions2));
```
Comment (baldersheim) [inline metrics-proxy/src/test/java/ai/vespa/metricsproxy/service/MetricsParserTest.java]: I added test and also removed the 2 System.out.println lines assuming they were debug.

**Review by baldersheim — CHANGES_REQUESTED** [review]
Either a set as suggested by @hakonhall or 
hash += Hasher.xx3(dim.id.getBytes(UTF8)) + Hasher.xx3(dim.value.getBytes(UTF8))

**Review by baldersheim — COMMENTED** [review]
Looking at the code and the history the key point was to detect unique Dimension sets.
And using a set seems a lot safer.

**Review by baldersheim — APPROVED** [review]

**Comment by gjoranv** [comment]
👍

File: metrics-proxy/src/main/java/ai/vespa/metricsproxy/service/MetricsParser.java [hunk 3]
```diff
@@ -142,25 +143,42 @@ static private void handleValue(JsonParser parser, Instant timestamp, Consumer c
...
             }
         }
-        Long uniqueKey = (((long) keyHash) << 32) | (valueHash & 0xffffffffL);
-        return uniqueDimensions.computeIfAbsent(uniqueKey, key -> dims.stream().collect(Collectors.toUnmodifiableMap(e -> toDimensionId(e.getKey()), Map.Entry::getValue)));
+        return uniqueDimensions.computeIfAbsent(dimensionsHashCode(dimensions),
+                                                key -> dimensions.stream().collect(Collectors.toUnmodifiableMap(
+                                                        dim -> toDimensionId(dim.id), dim -> dim.value)));
+    }
+
+    static long dimensionsHashCode(List<Dimension> dimensions) {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply metrics-proxy/src/main/java/ai/vespa/metricsproxy/service/MetricsParser.java]: Yes, it could

**Comment by baldersheim** [comment]
A benchmark after that change show parse time increased around 15-20% when using set. No effect on the memory footprint which is the most important measure here.

**Pull request merged**